File: src/48.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        //n*n 正方形矩阵 
        //做过类似的题，一圈一圈，由外向里
        //比较费时间，开始
        //今早烦的一批，就直接这个吧 
        int n = matrix.size();
        if(n == 1){
            return;
        }
        //int n = matrix.size();
        // C++ 这里的 = 拷贝是值拷贝，会得到一个新的数组
        auto matrix_new = matrix;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                matrix_new[j][n - i - 1] = matrix[i][j];
            }
        }
        // 这里也是值拷贝
        matrix = matrix_new;
        /*
        //计算每圈的起始点
        for(int i=0; i < (n/2); i++){
            int length =  (n - i * 2);
            //开始旋转一圈
            rotate1Circle(i,  length, n, matrix);
            //cout<<i<<","<<i<<","<<length<<endl;
        }
        return;
        */
    }
    /*
    void rotate1Circle(int _sPoint, int _side, int _n, vector<vector<int>>& _matrix){
        vector<int> tempSide(_side, 0); //(_matrix[_sPoint].begin(),  _matrix[_sPoint].end());
        //cout<<_sPoint<<","<<_side<<","<<_n<<endl;
        //上换右
        for(int i =_sPoint; i<_sPoint+_side;i++){
            //计算右侧的对应点位
            tempSide[i-_sPoint] = _matrix[i][_sPoint+_side-1];
            //赋值
            _matrix[i][_sPoint+_side-1] = tempSide[i-_sPoint];
            cout<<_matrix[i][_sPoint+_side-1]<<" "<<i<<"  "<<_sPoint+_side-1<<endl;
            //swap( _matrix[i][_sPoint+_side-1], tempSide[i-_sPoint]);
        }
        //右换下 
        for(int i=_sPoint;i<_sPoint+_side;i++){
            //计算下侧的点位 
            //_matrix[_sPoint+_side-1][_sPoint+_side-1+(i-_sPoint)]
            swap( _matrix[_sPoint+_side-1][_sPoint+_side-1-(i-_sPoint)], tempSide[i-_sPoint]);
        }
        //下换左
        for(int i=_sPoint+_side-1;i>_sPoint;i--){
            //计算左侧的对应点位
            //_matrix[_sPoint+_side-1-(i-_sPoint)][_sPoint]
            swap( _matrix[_sPoint+_side-1-(i-_sPoint)][_sPoint], tempSide[(_sPoint+_side-1)-i]);
        }
        //左换上
        for(int i=_sPoint+_side-1;i>_sPoint;i--){
            //计算上侧的对应点位
            //_matrix[_sPoint][_sPoint+(_sPoint+_side-1)-1]
            swap( _matrix[_sPoint][_sPoint+(_sPoint+_side-1)-1], tempSide[(_sPoint+_side-1)-i]);
        }
        */
    //}
};

File: src/48.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        #干翻微软！
        #先整个垃圾的，不是原地翻转的那种
        #矩阵起码包含一行一列
        #然后螺旋翻转就可以了
        #对每一个元素
        #好好做，看了题解，用上下翻转加对角线翻转
        #原地
        #时间击败-91.32%，空间击败-80.26%
        length = len(matrix)    #n*n矩阵，长宽相等
        #先上下翻转
        #奇数行中间那行就可以不翻转
        for row in range(length // 2):
            for col  in range(length):
                matrix[row][col],  matrix[length - row - 1][col] = matrix[length - row - 1][col],  matrix[row][col]
        #print(matrix)
        #再对角线翻转，把左下翻转到右上去
        #只能翻1/2的，要不都翻回来了
        for row in range(length):
            for col in range(row):
                matrix[row][col],  matrix[col][row] = matrix[col][row],  matrix[row][col]
        '''
        length = len(matrix)
        weight = len(matrix[0])
        tempMatrix = list()
        for i in range(length):
            aRow = list()
            for i in range(weight):
                aRow.append(0)
            tempMatrix.append(aRow)
        #print(tempMatrix)
        for i  in range(length):
            for j in range(weight):
                #print(i ,j, j,  length - i - 1)
                tempMatrix[j][length -i -1] = matrix[i][j]
        #print(tempMatrix)
        matrix[:] = tempMatrix
        '''
